Restore throttle user increment when add throttle user

devel-bobm opened the pull request:
Restores `AbstractThrottleManager` mechanism to count users of a given `Throttle`.

This is an attempt to address problem identified by @SteveRawlinson in #4873.

devel-bobm commented:
A quick check of java/src and jython directories does not uncover any use of the three-parameter version of `AbstractThrottleManager.notifyThrottleKnown()` except within `AbstractThrottleManager`, so that version is removed in favor of the (original) two-parameter version.

rhwood commented on java/src/jmri/jmrix/AbstractThrottleManager.java:
```diff
@@ -606,11 +596,14 @@ public boolean addressStillRequired(LocoAddress la) {
 
     @Override
     public void releaseThrottle(DccThrottle t, ThrottleListener l) {
+        log.debug("AbstractThrottleManager.releaseThrottle: {}, {}", t.toString(), l.toString());
```
We want to avoid using `toString()` in logging statements, since the String gets generated before determining if the logging statement should be written. The fastest way to (not) log in this case is `log.debug("AbstractThrottleManager.releaseThrottle: {}, {}", t, l);`.

devel-bobm replied:
fixed in subsequent commit by removing the .toString() invocations, as requested..

rhwood commented on java/src/jmri/jmrix/AbstractThrottleManager.java:
```diff
@@ -606,11 +596,14 @@ public boolean addressStillRequired(LocoAddress la) {
 
     @Override
     public boolean disposeThrottle(DccThrottle t, ThrottleListener l) {
+        log.debug("AbstractThrottleManager.disposeThrottle: {}, {}", t.toString(), l.toString());
```
Remove the `toString()` calls in this log line. In addition to the above concerns about `toString()` and logging, this call here forces the ThrottleListener (`l`) passed to be non-null, while the rest of this method accepts `l` being null. Its this line that is causing all CI tests to fail.

Alternatively, annotate (using `@Nonnull`) the ThrottleListener parameter as requiring a non-null ThrottleListener. Note that at least the `jmri.jmrix.can.cbus.CbusThrottleManager` expects to be able to call this method with a null ThrottleListener (and a number of tests do as well, even if the tested classes normally do not use null ThrottleListeners).

pabender commented on java/src/jmri/jmrix/AbstractThrottleManager.java:
```diff
@@ -830,4 +824,4 @@ boolean containsListener(ThrottleListener l) {
 
     private final static Logger log = LoggerFactory.getLogger(AbstractThrottleManager.class);
-}
+}
```
In addition to the things @rhwood mentioned, there is some strange character at the end of the file ( I can’t tell what it is from the web interface)

devel-bobm replied:
Lack of CRLF before EOF.  Added CRLF in subsequent commit.

devel-bobm commented:
Thanks for the reviews!  I've addressed the review comments - they all make sense to me, and have been addressed.

@pabender , the "funny character" issue apparantly was a missing CRLF at the EOF.  This is addressed in the latest commit.

pabender commented on java/test/jmri/jmrix/loconet/LnThrottleManagerTest.java:
```diff
@@ -20,9 +21,12 @@
 public class LnThrottleManagerTest extends jmri.managers.AbstractThrottleManagerTestBase {
 
     private DccThrottle throttle;
+    private DccThrottle throttle2;
```
Since this is only used in a single test, I think this should be a local variable in that test

devel-bobm replied:
I envision more testcases using multiple-throttle setups, so made the "second" copies "fields".  But there's no good reason why they cannot be test-specific local variables.

devel-bobm replied:
Actually, since these variables are assigned values within the listener, they need to be fields, not local variables.  (At least I can't get code to compile if I move the declarations into the tests.)

pabender commented on java/test/jmri/jmrix/loconet/LnThrottleManagerTest.java:
```diff
@@ -20,9 +21,12 @@
+    private DccThrottle throttle2;
 
     boolean failedThrottleRequest = false;
+    boolean failedThrottleRequest2 = false;
```
Since this is only used in a single test, I think this should be a local variable in that test

devel-bobm replied:
I envision more testcases using multiple-throttle setups, so made the "second" copies "fields".  But there's no good reason why they cannot be test-specific local variables.

devel-bobm replied:
Since these variables are assigned values within the listener, they need to be fields, not local variables.

pabender commented on java/test/jmri/jmrix/loconet/LnThrottleManagerTest.java:
```diff
@@ -20,9 +21,12 @@
     boolean failedThrottleRequest = false;
+    boolean failedThrottleRequest2 = false;
     int flagGotStealRequest = -1;
+    int flagGotStealRequest2 = -1;
```
Since this is only used in a single test, I think this should be a local variable in that test

devel-bobm replied:
Since these variables are assigned values within the listener, they need to be fields, not local variables.

pabender replied:
Ok, thanks for checking.

devel-bobm commented:
Appveyor failed on unrelated failure and is being restarted.  From the log:
```
    [junit] ------------- ---------------- ---------------
    [junit] Testcase: testRcvReplyShort(jmri.jmrix.tmcc.SerialTrafficControllerTest):	Caused an ERROR
    [junit] Unexpected ERROR or FATAL messages emitted
    [junit] java.lang.AssertionError: Unexpected ERROR or FATAL messages emitted
    [junit] 	at apps.tests.Log4JFixture.tearDown(Log4JFixture.java:38)
    [junit] 	at jmri.jmrix.tmcc.SerialTrafficControllerTest.tearDown(SerialTrafficControllerTest.java:274)
    [junit] 
    [junit] 
mvn :     [junit] Test apps.tests.AllTest FAILED
```

devel-bobm commented:
@SteveRawlinson,

In my view, `LocoNetThrottleManager` is doing exactly the right thing - letting `AbstractThrottleManager` determine whether or not the "concrete throttle object" (i.e. `LocoNetThrottle`) is needed.

`LnThrottleManager.releaseThrottle()` defers to `AbstractThrottleManager.releaseThrottle()`.  That invokes `LnThrottleManager.disposeThrottle()`.  `LnThrottleManager.disposeThrottle()` invokes `AbstractThrottleManager.disposeThrottle()`.  That invokes `AbstractThrottleManager.addressReleased()` to determine if the concrete throttle (i.e. `LocoNetThrottle` object or `NceThrottle` object or ... ) is still being used by some `jmri.Throttle` object.  If `AbstractThrottleManager.addressReleased()` says it is in use, then  `AbstractThrottleManager.disposeThrottle()` returns "false" to `LnThrottleManager.disposeThrottle()`, which sees the "false" and does nothing (leaving the `LocoNetThrottle` object intact and usable).  If instead `AbstractThrottleManager.addressReleased()` says there are no jmri.Throttle objects attached to the throttle, it returns false and `AbstractThrottleManager.disposeThrottle()` returns "true" to `LnThrottleManager.disposeThrottle()`, which sees the "true" and disposes the `LocoNetThrottle` object.

It seems to me that this is perfectly acceptable behavior for the LocoNet implementation.

**I believe that your specific concerns are NOT concerns with the LocoNet implementation** - they instead seem to be problems with how `AbstractThrottleManager.addressReleased()` determines whether the address is still needed.  _Please focus at that level, not at the LocoNet level!  That will solve your concerns across all system-types, not just LocoNet-based systems!_

As I do not have any knowledge of or experience with the mechanisms you use, which pay attention to a throttle without registering a ThrottleListener, I will have to defer to you to implement and test any changes to the `AbstractThrotleManager` behavior.

devel-bobm commented:
@SteveRawlinson ,

After thinking on this a little bit, I see that `AbstractThrottleManager` doesn't have direct visibility to the  concrete implementation of `AbstractThrottle` (i.e. the LocoNetTrottle, in this case), so it cannot know anything about that throttle's registered `PropertyChangeListeners`.

I will post a pull-request which modifies `LnThrottleManager.disposeThrottle()` to add sensitivity to the throttle's `PropertyChangeListeners` list, and abort the `disposeThrottle()` in cases where the list of listeners is not empty.

SteveRawlinson commented:
> Please focus at that level, not at the LocoNet level! That will solve your concerns across all system-types, not just LocoNet-based systems!

I'm reasonably confident this is a loconet specific issue. It hasn't happened for two years because the use count increment suppression has masked it, so it's unlikely you've seen it, but if you remove that suppression this problem will rise like Lazarus. 

AbstractThrottleManager sometimes returns a throttle object which as far as it knows is usable, but which has a slot with an inappropriate status and which is therefore useless.

The reason the slot gets in that state is that the loconet dispatchThrottle() method sets the slot status to COMMON without first checking whether there are any other users of the throttle object.

devel-bobm commented:
@SteveRawlinson 
The more I look into this set of problems, the more I think that there are non-trivial issues in how the `AbstractThrottle`, `AbstractThrottleManager` and the concrete implementations of those to classes, including the LocoNet implementation.

Please see #4897, and especially the comment which begins with "**BEWARE**".
